server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/ucacoin/exchanges/bkex/BkexExchange.java

```diff
@@ -0,0 +1,571 @@
+import java.lang.reflect.Method;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.net.URI;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.SSLContext;
+import javax.ws.rs.HeaderParam;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.generalbytes.batm.common.currencies.CryptoCurrency;
+import com.generalbytes.batm.server.extensions.IExchangeAdvanced;
+import com.generalbytes.batm.server.extensions.IRateSourceAdvanced;
+import com.generalbytes.batm.server.extensions.ITask;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.Balance;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.Balances;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.CreateOrder;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.CreatedOrder;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.DepositAddress;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.DepositAddressResponse;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.MarketTick;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.OrderBookSnapshot;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.dto.OrderState;
+import com.generalbytes.batm.server.extensions.util.net.CompatSSLSocketFactory;
+
+import si.mazi.rescu.ClientConfig;
+import si.mazi.rescu.RestProxyFactory;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.RateLimiter;
+
+public class BkexExchange implements IExchangeAdvanced, IRateSourceAdvanced {
+
+    private static final Logger LOG = LoggerFactory.getLogger("batm.master.bkexExchange");
+
+    private static final URI LIVE_URL = URI.create("https://api.bkex.io");
+
+    public static IRateSourceAdvanced asRateSource(String preferredFiatCurrency) {
+        return new BkexExchange(LIVE_URL, null, null, preferredFiatCurrency);
+    }
+
+    public static IExchangeAdvanced asExchange(String apikey, String secret, String preferredFiatCurrency) {
+        Objects.requireNonNull(apikey, "apikey");
+        Objects.requireNonNull(apikey, "secret");
+
+        return new BkexExchange(LIVE_URL, apikey, secret, preferredFiatCurrency);
+    }
+
+    enum TradableLimitFor {
+        BUY, SELL
+    }
+
+    // all CRYPTO_FIAT combinations are supported
+    private static final ImmutableSet<String> FIAT_CURRENCIES = ImmutableSet.of(CryptoCurrency.USDT.getCode());
```

Comment:
USDT cannot be fiat currency. Use USD in this set and later when calling exchange remap it to USDT.